cohortextractor2/query_engines/base_sql.py

```diff
@@ -58,14 +58,8 @@ class BaseSQLQueryEngine(BaseQueryEngine):
     # No limit by default although some DBMSs may impose one
     max_rows_per_insert: Optional[int] = None
 
-    def __init__(self, column_definitions, backend):
-        super().__init__(column_definitions, backend)
-        self._engine = None
-
-        # Walk the nodes and identify output groups
-        all_nodes = self.get_all_query_nodes(column_definitions)
-        self.output_groups = self.get_output_groups(all_nodes)
-        self.codelists = [node for node in all_nodes if isinstance(node, Codelist)]
+    # Per-instance cache for SQLAlchemy Engine
+    _engine = None
```

Comment:
Yes, and no :) Because instance attributes shadow class attributes (and because `None` is an immutable value) the net effect is that it acts as instance default, meaning `self._engine` is `None` until you set it.

That feels like a fairly widespread Python idiom to me, but I can add a comment if you think that would be helpful.

Reply:
No need! If it's a widespread idiom then I shall internalize it. Thanks.